// webOptimizer/pugLoader.compiled.js
#!/usr/bin/env node

// -*- coding: utf-8 -*-
'use strict';
/* !
    region header
    Copyright Torben Sickert (info["~at~"]torben.website) 16.12.2012

    License
    -------

    This library written by Torben Sickert stand under a creative commons naming
    3.0 unported license. see http://creativecommons.org/licenses/by/3.0/deed.de
    endregion
*/
// region imports

var _fs = require('fs');

var fileSystem = _interopRequireWildcard(_fs);

var _pug = require('pug');

var pug = _interopRequireWildcard(_pug);

var _loaderUtils = require('loader-utils');

var loaderUtils = _interopRequireWildcard(_loaderUtils);

var _helper = require('./helper.compiled');

var _helper2 = _interopRequireDefault(_helper);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj.default = obj; return newObj; } }

// NOTE: Only needed for debugging this file.
try {
    require('source-map-support/register');
} catch (error) {}

// endregion

// endregion
// region types
module.exports = function (source) {
    if (this.cacheable) this.cacheable();
    const query = _helper2.default.extendObject(true, {
        moduleAliases: [], knownExtensions: ['.pug', '.html', '.js', '.css'],
        context: './'
    }, this.options.pug || {}, loaderUtils.parseQuery(this.query));
    const compile = (template, options = query.compiler) => (locals = {}) => {
        options = _helper2.default.extendObject(true, {
            filename: template, doctype: 'html',
            compileDebug: this.debug || false
        }, options);
        let templateFunction;
        if (options.isString) {
            delete options.isString;
            templateFunction = pug.compile(template, options);
        } else templateFunction = pug.compileFile(template, options);
        return templateFunction(_helper2.default.extendObject(true, {
            require: request => {
                const template = request.replace(/^(.+)\?[^?]+$/, '$1');
                const queryMatch = request.match(/^.+\?([^?]+)$/, '$1');
                let nestedLocals = {};
                if (queryMatch) {
                    const evaluationFunction = (request, template, source, compile, locals) =>
                    // IgnoreTypeCheck
                    new Function('request', 'template', 'source', 'compile', 'locals', `return ${ queryMatch[1] }`)(request, template, source, compile, locals);
                    nestedLocals = evaluationFunction(request, template, source, compile, locals);
                }
                const options = _helper2.default.extendObject(true, {
                    encoding: 'utf-8'
                }, nestedLocals.options || {});
                if (options.isString) return compile(template, options)(nestedLocals);
                const templateFilePath = _helper2.default.determineModuleFilePath(template, query.moduleAliases, query.knownExtensions, query.context);
                this.addDependency(templateFilePath);
                if (queryMatch || templateFilePath.endsWith('.pug')) return compile(templateFilePath, options)(nestedLocals);
                return fileSystem.readFileSync(templateFilePath, options);
            } }, locals));
    };
    return compile(source, _helper2.default.extendObject(true, {
        isString: true,
        filename: loaderUtils.getRemainingRequest(this).replace(/^!/, '')
    }, query.compiler || {}))(query.locals || {});
};
// region vim modline
// vim: set tabstop=4 shiftwidth=4 expandtab:
// vim: foldmethod=marker foldmarker=region,endregion:
// endregion
